### ISSUE TITLE
FetchedRecordsController index path sections

### ISSUE THREAD

**Issue by palpatim** [groue/GRDB.swift#86]
Hi Gwendal,

I'm attempting to implement a FetchedRecordsController for a UICollectionView with multiple sections, and figuring out how to translate the index paths. The docs for FetchedRecordsController say:

> ☝️ Note: In its current state, FetchedRecordsController does not support grouping table view rows into custom sections: it generates a unique section.

Can I interpret that to mean:

* The returned indexPath.section on a tableViewEvent is always guaranteed to be 0
* `fetchedRecords[indexPath.row]` references the same item as specified by the event's `indexPath`

Basically, I'm trying to map my sections so that if I have a collection view, backed by a single FetchedRecordsController:

```
section0: [item0, item1, item2]
section1: [item3]
section2: [item4, item5, item6, item7]
```

Then I could translate an .Insertion event for indexPath (0, 6) as referring to `item6`, or indexPath (2, 2).

Finally, if I have a collectionView with hundreds of sections, I would imagine it to be inefficient to set up a FetchedRecordsController for each section. Is that correct, or are they lightweight enough to make that a viable strategy?

**Comment by groue** [comment]
Hello Tim,

> Can I interpret that to mean:
> 
> * The returned indexPath.section on a tableViewEvent is always guaranteed to be 0
> * `fetchedRecords[indexPath.row]` references the same item as specified by the event's `indexPath`

Yes and yes. Specifically, the assertion below holds:

```swift
controller.trackChanges(
    tableViewEvent: { (controller, record, event) in
        switch event {
        case .Insertion(let indexPath):
            assert(controller.fetchedRecords![indexPath.row].id == record.id)
            ...
```

> Basically, I'm trying to map my sections so that if I have a collection view, backed by a single FetchedRecordsController

I don't really see how you will issue *section* insertion/deletion/move events from individual *item* events, surely a non-trivial task, but let's assume that you known what you are doing.

Also, I must say that FetchedRecordsController has been tested with UITableView, but not UICollectionView. UITableView is very sensitive to events ordering, and I can just hope that UICollectionView is compatible.

Just to be clear: if you don't need UICollectionView *animations*, things are much simpler since you'd just have to map index paths and issue `reloadData` in the `recordsDidChange` callback.

> Finally, if I have a collectionView with hundreds of sections, I would imagine it to be inefficient to set up a FetchedRecordsController for each section. Is that correct, or are they lightweight enough to make that a viable strategy?

No, they are not lightweight enough. Each FetchedRecordsController will run its fetch request as soon as there is a change in the involved columns of the involved tables. That would mean hundreds of SQLite SELECT for each transaction that contains an insertion, update, or deletion.

I hope to bring multiple sections support to FetchedRecordsController in the future, but this has not been done yet.

**Comment by palpatim** [comment]
Thank you for the confirmation and explanation. I'll be interested to see your section implementation!
